sympy/diffgeom/diffgeom.py

```diff
@@ -1084,6 +1086,388 @@ def __call__(self, field):
+        the 2nd kind as argument.
+        """
+        ch2_tensor = TensorArray(components=ch2,variance=[-1,1,1],coordinate_system=self.CoordSystem)
+        ch2Xself = ch2_tensor.TensorProduct(self)
+        contractions = self.zero([k+1 for k in self.contravariant_slots],[0]+[k+1 for k in self.covariant_slots])
+        for k in self.covariant_slots:
+            newcontraction = ch2Xself.contract(0,k+3)
+            newcontraction._move_index(1,k+1)
+            assert(newcontraction.covariant_slots == contractions.covariant_slots)
+            assert(newcontraction.tensor.shape == contractions.tensor.shape)
+            contractions = contractions - newcontraction
+        for k in self.contravariant_slots:
+            newcontraction = ch2Xself.contract(1,k+3)
+            newcontraction._move_index(0,k+1)
+            assert(newcontraction.covariant_slots == contractions.covariant_slots)
+            contractions = contractions + newcontraction
+        coord_derivative = self.zero([k+1 for k in self.contravariant_slots],[0]+[k+1 for k in self.covariant_slots])
+        for index in self.indices:
+            for k in range(self.n):
+                coord_derivative.tensor[(k,)+index] = self.base_vectors[k](self.tensor[index])
+        return contractions + coord_derivative
+
+class ToArray(TensorArray):
+    """
+    ``ToArray(T)``: if T is a TensorProduct, or obtained by applying a CovD
+    or other tensor operation, this will represent T as a TensorArray,
+    allowing array operations (such as contractions and braiding) to be
+    performed on the tensor.  It can then be converted back into a tensor
+    product object with the TensorArray.to_tensor method.
+
+    Example 1: Basic usage to expand tensors
+    ========================================
+
+    >>> from sympy.diffgeom import WedgeProduct, ToArray, Differential
+    >>> from sympy.diffgeom.rn import R2_r
+    >>> wp=WedgeProduct(Differential(R2_r.x),Differential(R2_r.y))
+    >>> ToArray(wp).to_tensor()
+    TensorProduct(dx, dy) - TensorProduct(dy, dx)
+
+    Example 2: Covariant derivative as a tensor
+    ===========================================
+    >>> from sympy.diffgeom import (Manifold, Patch, CoordSystem, TensorProduct, metric_to_Christoffel_2nd, ToArray)
+    >>> from sympy import (Function, sin)
+    >>> TP = TensorProduct
+    >>> M = Manifold('Reisner-Nordstrom', 4)
+    >>> p = Patch('origin', M)
+    >>> cs = CoordSystem('spherical', p, ['t', 'r', 'theta', 'phi'])
+    >>> t, r, theta, phi = cs.coord_functions()
+    >>> dt, dr, dtheta, dphi = cs.base_oneforms()
+    >>> f=Function('f')
+    >>> metric = f(r)**2*TP(dt, dt) - f(r)**(-2)*TP(dr, dr) - r**2*TP(dtheta, dtheta) - r**2*sin(theta)**2*TP(dphi, dphi)
+    >>> ch_2nd = metric_to_Christoffel_2nd(metric)
+    >>> G=ToArray(metric)
+    >>> assert(G.covD(ch_2nd).to_tensor() == 0)  #the covariant derivative as a (3,0) tensor!
+    """
+    def __init__(self,T, **kwargs):
```

Comment:
In SymPy we usually try to have the same parameters of the superclass in the constructor of a subclass. Maybe adding some to the end.